### PR TITLE
Improvements to the language editor

### LINKED ISSUE #4855: Language editor can't handle values with ) correctly with array values
If using the inbuilt language editor to edit an array, like the months array in index.english.php, the values currently must not have ) in them, as these are used for string matching.

Arbitrary example:
```
$txt['months'] = array(1 => 'January(Jan) Jan', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December');
```

Renders in the editor for `months-+- 1` as `January(JanJan`

Also, it can't handle string indexes which will make life interesting supporting the proposed fixes for #2020.

### LINKED ISSUE #4856: Language editor truncates weirdly on Help
Trying to use the inbuilt language editor on the Help language strings results in truncation of the editable string. Not sure why yet.

image

### PR DESCRIPTION
**Pull request by Sesquipedalian**
1. Fixes #4855
2. Fixes #4856 
3. Adds language editor support for literal string keys in $txt arrays.
4. Adds language editor support for `$txt['foo']['bar'] = 'baz';` syntax.
5. Adds support for adding and removing values inside $txt arrays.
6. Gracefully handles trailing comments and whitespace in language files.
7. Tightens up some logic and eliminates some code duplication.
8. Handles $txt arrays as proper arrays on HTML form.
9. Makes some minor improvements to the language editor UI.
10. Adds `native_name` and `lang_recaptcha` to the primary settings of a language.
11. Allows mods to add additional string types via integration hook.
12. Fixes various unreported bugs discovered during testing.

### CODE REVIEW

**Comment by Sesquipedalian** [comment]
To test, try pasting this into Modifications.english.php and then messing around with the values in the language editor:

```
$txt['mline'] = 'This has multiple lines in it.
It is followed by int_elements[0].
This line ends in a semicolon;
Does this line show up too?';

$txt['int_elements'][0] = 'foo';
$txt['int_elements'][1] = 'bar';
$txt['int_elements'][2] = 'baz';

$txt['string_elements']['a'] = 'foo';
$txt['string_elements']['b'] = 'bar';
$txt['string_elements']['c'] = 'baz';

$txt['string_key_array'] = array('a' => 'foo', 'b' => 'bar', 'c' => 'baz');

$txt['int_key_array'] = array(1 => 'foo', 'bar', 12 => 'baz');

$txt['simple_array'] = array('foo', 'bar', 'baz');

$txt['simple_string_foo'] = 'foo';
$txt['simple_string_bar'] = 'bar';
$txt['simple_string_baz'] = 'baz';
```

File: Themes/default/languages/ManageSettings.english.php [hunk 1]
```diff
@@ -202,6 +202,7 @@
 $txt['recaptcha_secret_key_desc'] = 'This is for communication between your site and Google. Be sure to keep it a secret.';
 $txt['recaptcha_no_key_question'] = 'Don\'t have a key for reCAPTCHA?';
 $txt['recaptcha_get_key'] = 'Get your reCAPTCHA key here.';
+$txt['languages_recaptcha'] = '<a href="https://developers.google.com/recaptcha/docs/language" target="_blank" class="bbc_link">ReCAPTCHA language</a>';
```
Comment (frandominguezl) [inline Themes/default/languages/ManageSettings.english.php]: Can't this be formatted in other way?

Comment (Sesquipedalian) [reply Themes/default/languages/ManageSettings.english.php]: Oh, fine, if you're going to be all picky about it, I'll make a $helptxt entry for it. 😜

Comment (frandominguezl) [reply Themes/default/languages/ManageSettings.english.php]: d3vcho now rests peacefully *

**Comment by Sesquipedalian** [comment]
I decided to change again the method for extracting the individual $txt variables from the file. Even though the method I was using as of 095e277 seems to have worked well, using `file_get_contents()` and `preg_split()` to do the job (1) is much simpler, (2) eliminates any possibility of problems that could arise from processing lines one at a time (the root cause of most of the bugs), and (3) performs fewer total preg_* operations to do the job.

**Comment by Sesquipedalian** [comment]
Unless anyone finds any bugs, this is done.

**Comment by jdarwood007** [comment]
If the file is empty, doesn't seem to let you add entries.

$ cat Themes/default/languages/Modifications.english.php
```
<?php
// Version: 2.1 Beta 4; Modifications

?>
```

Try to add a entry and it saves but nothing is added.

**Comment by Sesquipedalian** [comment]
That should be fixed now, @jdarwood007

**Review by jdarwood007 — APPROVED** [review]

**Comment by jdarwood007** [comment]
Looks good to me now.  Not sure if for a commit of this size if we should have another reviewer.

File: Sources/ManageLanguages.php [hunk 2]
```diff
@@ -1233,21 +1401,66 @@ function ModifyLanguage()
...
+						if (bracket_regex.test(key)) {
+							alert("' . $txt['languages_invalid_key'] . '" + key + subkey);
+							return;
+						}
+
+						$("#language_" + group).append("<dt><span>" + key + subkey + "</span></dt> <dd id=\"entry_" + entry_num + "\"><input id=\"entry_" + entry_num + "_edit\" class=\"entry_toggle\" type=\"checkbox\" name=\"edit[" + key + "]" + subkey + "\" value=\"add\" data-target=\"#entry_" + entry_num + "\" checked> <label for=\"entry_" + entry_num + "_edit\">' . $txt['edit'] . '</label> <input type=\"hidden\" class=\"entry_oldvalue\" name=\"grp[" + key + "]\" value=\"" + group + "\"> <textarea name=\"entry[" + key + "]" + subkey + "\" class=\"entry_textfield\" cols=\"40\" rows=\"1\" style=\"width: 96%; margin-bottom: 2em;\"></textarea></dd>");
+					}
+				};');
+
+			addInlineJavaScript('
```
Comment (albertlast) [inline Sources/ManageLanguages.php:1459]: Is ther a reason why do two addinlinejs calls instead of one?

Comment (Sesquipedalian) [reply Sources/ManageLanguages.php:1459]: No, there is not.

Comment (Sesquipedalian) [reply Sources/ManageLanguages.php:1459]: Oh, yes there is. One is deferred and the other is not.

**Pull request merged**